cmake/modules/extensions.cmake

```diff
@@ -4786,3 +4787,138 @@ macro(zephyr_check_flags_exclusive function prefix)
       )
   endif()
 endmacro()
+
+########################################################
+# 7. Linkable loadable extensions (llext)
+########################################################
+#
+# These functions simplify the creation and management of linkable
+# loadable extensions (llexts).
+#
+
+# Add a custom target that compiles a single source file to a .llext file.
+#
+# Output and source files must be specified using the OUTPUT and SOURCES
+# arguments. Only one source file is currently supported.
+#
+# The llext code will be compiled with mostly the same C compiler flags used
+# in the Zephyr build, but with some important modifications. The list of
+# flags to remove and flags to append is controlled respectively by the
+# LLEXT_REMOVE_FLAGS and LLEXT_APPEND_FLAGS global variables.
+
+# The C_FLAGS argument can be used to pass additional compiler flags to the
+# compilation of this particular llext.
+#
+# Example usage:
+#   add_llext_target(hello_world
+#     OUTPUT  ${PROJECT_BINARY_DIR}/hello_world.llext
+#     SOURCES ${PROJECT_SOURCE_DIR}/src/llext/hello_world.c
+#     C_FLAGS -Werror
+#   )
+# will compile the source file src/llext/hello_world.c to a file
+# ${PROJECT_BINARY_DIR}/hello_world.llext, adding -Werror to the compilation.
+#
+function(add_llext_target target_name)
+  set(single_args OUTPUT)
+  set(multi_args SOURCES;C_FLAGS)
+  cmake_parse_arguments(PARSE_ARGV 1 LLEXT "${options}" "${single_args}" "${multi_args}")
+
+  # Check that the llext subsystem is enabled for this build
+  if (NOT CONFIG_LLEXT)
+    message(FATAL_ERROR "add_llext_target: CONFIG_LLEXT must be enabled")
+  endif()
```

Comment:
Just curious, what would be wrong with trying/wanting to generate an llext for a project that doesn't have LLEXT enabled. One needn't *use* LLEXT in a project to be able to have "snippets" of it built as llext, no?

Reply:
Well, indeed you could compile only the llext, since there's no real technical barrier. However, the generated file would not work with the same Zephyr core with which it was built (no exported symbols or machinery to load it), so it looked like a config error in my eyes.

---

Now that I look at it again, the converse is useful though - if you know what you are doing and _purposefully disable_ `CONFIG_LLEXT` on a project that defines such targets, it makes sense not to build them _and not complain_, unless they are required by other parts in the project.

Taking into account your other comment on dependencies ("why isn't a llext built if it's defined as part of a project?"), I think the most appropriate way to deal with all this should be:
* if `CONFIG_LLEXT` is not enabled, do not complain in `add_llext_target` and define an empty target instead, so other `llext_`* APIs don't break as well.
* automatically add the llext target to ALL so it's always built if it's defined in the CMake.

This would result in an actual build error only when LLEXT is disabled and the output file is a required dependency (for example by `generate_inc_file_for_target`) - or if the Zephyr code actually calls llext APIs, obviously.

@teburd, what do you think about this?

Reply:
Sorry for the belated reply and - much worse - sorry for not being @teburd.

> disable CONFIG_LLEXT on a project that defines such targets, it makes sense not to build them and not complain,

I probably (!) disagree with that? The main purpose of Kconfig is to turn software features on and off. But "llext" isn't really a software _feature_, it's more about HOW you want some of those features to be delivered: as a separate module or built-in? So CONFIG_LLEXT is more like a "dependency" of real features.

Consider the Linux kernel, which many `llext` aspects take inspiration from. It had to already deal with this problem. If you turn off Linux CONFIG_MODULES in `make menuconfig` then it immediately forces you to make all features built-in. Editing the .config and running `make oldefconfig` does the same thing.

Of course there's a massive difference: the Linux kernel can very quickly easily flip drivers that support between module and built-in, whereas `llext` does not do that all. But "inspiration" is still useful IMHO.

So if you:
- enable some particular llext feature in Kconfig, AND
- you turn off CONFIG_LLEXT

... then you have a **configuration error** and the build should 1) fail, 2) fail as soon as possible, not wait until compilation whether some symbol is used maybe, maybe not.

In other words, if you turn off CONFIG_LLEXT then you MUST also turn off all features that depend on it - and maybe Kconfig can/should help with that but either way the build should fail ASAP if you give it an inconsistent configuration.

Reply:
> I'll try to draft one shortly and share it so we can all agree on the long term path

Thank you so much for leading here again. Whatever are the use cases and intended Kconfig interactions, I think we all agree they cannot just be inferred from scattered comments in source and code reviews. Some high-level design documentation is required indeed.